[mlir] Use `OpBuilder::createBlock` in op builders and patterns

matthias-springer opened the pull request:
When creating a new block in (conversion) rewrite patterns, `OpBuilder::createBlock` must be used. Otherwise, no `notifyBlockInserted` notification is sent to the listener.

Note: The dialect conversion relies on listener notifications to keep track of IR modifications. Creating blocks without the builder API can lead to memory leaks during rollback.

ftynse commented on mlir/lib/Dialect/Linalg/Utils/Utils.cpp:
```diff
@@ -275,14 +275,13 @@ GenericOp makeTransposeOp(OpBuilder &b, Location loc, Value inputTensor,
-  b.create<YieldOp>(loc, transposeOp.getRegion().front().getArgument(0));
+  Region &body = transposeOp.getRegion();
+  Block *bodyBlock = b.createBlock(&body);
+  bodyBlock->addArguments({elementType, elementType}, {loc, loc});
```
Shouldn't this go through the rewriter (IIRC, there is an overload of `createBlock` that takes a list of arguments).

ftynse commented on mlir/lib/Dialect/Shape/IR/Shape.cpp:
```diff
@@ -1904,23 +1902,23 @@ bool ToExtentTensorOp::areCastCompatible(TypeRange inputs, TypeRange outputs) {
-  Block &bodyBlock = bodyRegion->front();
-  bodyBlock.addArgument(builder.getIndexType(), result.location);
+  Block *bodyBlock = builder.createBlock(bodyRegion);
+  bodyBlock->addArgument(builder.getIndexType(), result.location);
```
Ditto.

ftynse commented on mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorRewriting.cpp:
```diff
@@ -299,8 +299,7 @@ struct FuseSparseMultiplyOverAdd : public OpRewritePattern<GenericOp> {
+    Block *fusedBlock = rewriter.createBlock(&fusedOp.getRegion());
     unsigned num = prodBlock.getNumArguments();
     for (unsigned i = 0; i < num - 1; i++)
       addArg(mapper, fusedBlock, prodBlock.getArgument(i));
```
I suspect this may be also unhappy if the underlying implementation adds arguments to fusedBlock.

matthias-springer replied:
We have no `notifyBlockModified` callback at the moment, so I think there is no expectation that a listener be notified about adding/removing/changing block arguments. I have plans to add `RewriterBase::Listener::notifyBlockModified` and generalize `modifyOpInPlace` to `modifyInPlace` (accepting both ops and blocks).

matthias-springer committed:
[mlir] Use `OpBuilder::createBlock` in op builders and patterns

When creating a new block in (conversion) rewrite patterns, `OpBuilder::createBlock` must be used. Otherwise, no `notifyBlockInserted` notification is sent to the listener.

Note: The dialect conversion relies on listener notifications to keep track of IR modifications. Creating blocks without the builder API can lead to memory leaks during rollback.

joker-eph commented on mlir/lib/Conversion/MemRefToLLVM/MemRefToLLVM.cpp:
```diff
@@ -520,9 +520,7 @@ struct GlobalMemrefOpLowering
         global, arrayTy, global.getConstant(), linkage, global.getSymName(),
         initialValue, alignment, *addressSpace);
     if (!global.isExternal() && global.isUninitialized()) {
-      Block *blk = new Block();
```
Oh these raw uses of `new` always irked me deeply! Glad to see more that are gone

aartbik reviewed (commented):
lgtm
